crossdock/src/helpers.js

```diff
@@ -20,10 +20,196 @@
+        if (transport === constants.TRANSPORT_HTTP) {
+            return this.callDownstreamHTTP(downstream, context);
+        } else if (transport === constants.TRANSPORT_TCHANNEL) {
+            return this.callDownstreamTChannel(downstream, context);
+        } else if (transport == constants.TRANSPORT_DUMMY) {
+            return new RSVP.Promise((resolve, reject) => {
+                resolve({ 'notImplementedError': 'Dummy has not been implemented' });
+            });
+        } else {
+            return new RSVP.Promise((resolve, reject) => {
+                resolve({ 'notImplementedError': `Unrecognized transport received: ${transport}` })
+            });
+        }
+    }
+
+    callDownstreamHTTP(downstream: Downstream, context: any): any {
+        return new RSVP.Promise((resolve, reject) => {
+
+            // $FlowIgnore - Honestly don't know why flow compalins about family.
+            dns.lookup(downstream.host, (err, address, family) => {
+                if (err) {
+                    console.log('dns_lookup_err', err);
+                    return;
+                }
+
+                let port = parseInt(downstream.port);
+                let downstreamUrl = `http:\/\/${address}:${port}/join_trace`;
+
+                let clientSpan = this._tracer.startSpan('client-span', { childOf: context.span.context() });
+                let headers = { 'Content-Type': 'application/json' };
+                this._tracer.inject(clientSpan.context(), opentracing.FORMAT_HTTP_HEADERS, headers);
+
+                request.post({
+                    'url': downstreamUrl,
+                    'forever': true,
+                    'headers': headers,
+                    'body': JSON.stringify({
+                        'serverRole': downstream.serverRole,
+                        'downstream': downstream.downstream
+                    })
+                }, (err, response) => {
+                    if (err) {
+                        console.log('error in downstream call:', err);
+                        reject(err);
+                        clientSpan.finish();
+                        return;
+                    }
+
+                    clientSpan.finish();
+                    let downstreamResponse = JSON.parse(response.body);
+                    resolve(downstreamResponse);
+                });
+
+            });
+        });
+    }
+
+    callDownstreamTChannel(downstream: Downstream, context: any): any {
+        return new RSVP.Promise((resolve, reject) => {
+            dns.lookup(downstream.host, (err, address, family) => {
+                if (err) {
+                    console.log('dns_lookup_err', err);
+                    return;
+                }
+
+                let port = parseInt(downstream.port);
+                let downstreamUrl = `http:\/\/${address}:${port}/join_trace`;
+
+                let clientSpan = this._tracer.startSpan('client-span', { childOf: context.span.context() });
```

Comment:
if you are using wrapped/decorated tchannel, it should already start client span automatically